## Project 3
This project's objective is to clean data provided by the 'autos'.csv (Used Car Listing + Info) file and make it ready for analysis. 

In [1]:
import pandas as pd
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')


In [2]:
autos

dateCrawled                                               name  \
0      3/26/2016 17:47                   Peugeot_807_160_NAVTECH_ON_BOARD   
1       4/4/2016 13:38         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      3/26/2016 18:57                         Volkswagen_Golf_1.6_United   
3      3/12/2016 16:58  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4       4/1/2016 14:38  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                ...                                                ...   
49995  3/27/2016 14:38   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  3/28/2016 10:50  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997   4/2/2016 14:44                     Fiat_500_C_1.2_Dualogic_Lounge   
49998   3/8/2016 19:25                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999   3/14/2016 0:42                                Opel_Vectra_1.6_16V   

       seller offerType     price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000   control         bus                2004   
1      privat   Angebot   $8,500   control   limousine                1997   
2      privat   Angebot   $8,990      test   limousine                2009   
3      privat   Angebot   $4,350   control  kleinwagen                2007   
4      privat   Angebot   $1,350      test       kombi                2003   
...       ...       ...       ...      ...         ...                 ...   
49995  privat   Angebot  $24,900   control   limousine                2011   
49996  privat   Angebot   $1,980   control      cabrio                1996   
49997  privat   Angebot  $13,200      test      cabrio                2014   
49998  privat   Angebot  $22,900   control       kombi                2013   
49999  privat   Angebot   $1,250   control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage     dateCreated  nrOfPictures  postalCode  \
0         peugeot              nein  3/26/2016 0:00             0       79588   
1             bmw              nein   4/4/2016 0:00             0       71034   
2      volkswagen              nein  3/26/2016 0:00             0       35394   
3           smart              nein  3/12/2016 0:00             0       33729   
4            ford              nein   4/1/2016 0:00             0       39218   
...           ...               ...             ...           ...         ...   
49995        audi              nein  3/27/2016 0:00             0       82131   
49996        opel              nein  3/28/2016 0:00             0       44807   
49997        fiat              nein   4/2/2016 0:00             0       73430   
49998        audi              nein   3/8/2016 0:00             0       35683   
49999        opel              nein  3/13/2016 0:00             0       45897   

             lastSeen  
0       4/6/2016 6:45  
1      4/6/2016 14:45  
2      4/6/2016 20:15  
3      3/15/2016 3:16  
4      4/1/2016 14:38  
...               ...  
49995  4/1/2016 13:47  
49996  4/2/2016 14:18  
49997  4/4/2016 11:47  

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  seller  \
0  3/26/2016 17:47                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1   4/4/2016 13:38         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  3/26/2016 18:57                         Volkswagen_Golf_1.6_United  privat   
3  3/12/2016 16:58  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4   4/1/2016 14:38  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offerType    price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000   control         bus                2004    manuell   
1   Angebot  $8,500   control   limousine                1997  automatik   
2   Angebot  $8,990      test   limousine                2009    manuell   
3   Angebot  $4,350   control  kleinwagen                2007  automatik   
4   Angebot  $1,350      test       kombi                2003    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   
3       71  fortwo   70,000km                    6   benzin       smart   
4        0   focus  150,000km                    7   benzin        ford   

  notRepairedDamage     dateCreated  nrOfPictures  postalCode        lastSeen  
0              nein  3/26/2016 0:00             0       79588   4/6/2016 6:45  
1              nein   4/4/2016 0:00             0       71034  4/6/2016 14:45  
2              nein  3/26/2016 0:00             0       35394  4/6/2016 20:15  
3              nein  3/12/2016 0:00             0       33729  3/15/2016 3:16  
4              nein   4/1/2016 0:00             0       39218  4/1/2016 14:38

We observe that there are few columns that have nulls and that the columns are in Camel Case and not Snake Case

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [6]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_PS            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

date_crawled                                               name  seller  \
0  3/26/2016 17:47                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1   4/4/2016 13:38         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  3/26/2016 18:57                         Volkswagen_Golf_1.6_United  privat   
3  3/12/2016 16:58  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4   4/1/2016 14:38  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type    price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000   control          bus               2004    manuell   
1    Angebot  $8,500   control    limousine               1997  automatik   
2    Angebot  $8,990      test    limousine               2009    manuell   
3    Angebot  $4,350   control   kleinwagen               2007  automatik   
4    Angebot  $1,350      test        kombi               2003    manuell   

   power_PS   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   
3        71  fortwo   70,000km                   6    benzin       smart   
4         0   focus  150,000km                   7    benzin        ford   

  unrepaired_damage      ad_created  nr_of_pictures  postal_code  \
0              nein  3/26/2016 0:00               0        79588   
1              nein   4/4/2016 0:00               0        71034   
2              nein  3/26/2016 0:00               0        35394   
3              nein  3/12/2016 0:00               0        33729   
4              nein   4/1/2016 0:00               0        39218   

        last_seen  
0   4/6/2016 6:45  
1  4/6/2016 14:45  
2  4/6/2016 20:15  
3  3/15/2016 3:16  
4  4/1/2016 14:38

Analyze the Autos Dataframe by using 'describe()' which would indicate which series to adjust.

In [7]:
autos.describe()

registration_year      power_PS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [8]:
autos.describe(include = 'all')


date_crawled         name  seller offer_type  price abtest  \
count             50000        50000   50000      50000  50000  50000   
unique            12073        38754       2          2   2357      2   
top     3/28/2016 14:49  Ford_Fiesta  privat    Angebot    $0    test   
freq                 15           78   49999      49999   1421  25756   
mean                NaN          NaN     NaN        NaN    NaN    NaN   
std                 NaN          NaN     NaN        NaN    NaN    NaN   
min                 NaN          NaN     NaN        NaN    NaN    NaN   
25%                 NaN          NaN     NaN        NaN    NaN    NaN   
50%                 NaN          NaN     NaN        NaN    NaN    NaN   
75%                 NaN          NaN     NaN        NaN    NaN    NaN   
max                 NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

           ad_created  nr_of_pictures   postal_code      last_seen  
count           50000         50000.0  50000.000000          50000  
unique             76             NaN           NaN          10432  
top     4/3/2016 0:00             NaN           NaN  4/6/2016 2:45  
freq             1946             NaN           NaN            112  
mean              NaN             0.0  50813.627300            NaN  
std               NaN             0.0  25779.747957            NaN  
min               NaN             0.0   1067.000000            NaN  
25%               NaN             0.0  30451.000000            NaN  
50%               NaN             0.0  49577.000000            NaN  
75%               NaN             0.0  71540.000000            NaN  
max               NaN             0.0  99998.000000            NaN

We would need to make the 'price' and 'odometer' series into 'int', to do this we must replace all of the strings before we convert the string into an int. 

In [9]:
autos.loc[:, 'price'] = autos.loc[:, 'price'].str.replace(',','').str.replace('$', '').astype(int)

In [10]:
autos.loc[:, 'odometer'] = autos.loc[:, 'odometer'].str.replace('km', '').str.replace(',',  '').astype(int)

After we have done this, we use the info() function to check whether or not 'price' and 'odometer' has changed from an object to a string. 

In [11]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_PS            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  int32 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

We should rename the 'odometer' column label to 'odometer_km' since we replaced the 'km' to make this column an int.

In [12]:
autos.rename({'odometer': 'odometer_km'}, axis = 1 ,inplace = True)

In [13]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_PS            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int32 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

We must use .value_counts() and .describe() to see which column could be dropped. 

In [14]:
print(autos.describe(include = 'all'))
autos.loc[:, 'nr_of_pictures'].value_counts()

           date_crawled         name  seller offer_type         price abtest  \
count             50000        50000   50000      50000  5.000000e+04  50000   
unique            12073        38754       2          2           NaN      2   
top     3/28/2016 14:49  Ford_Fiesta  privat    Angebot           NaN   test   
freq                 15           78   49999      49999           NaN  25756   
mean                NaN          NaN     NaN        NaN  9.840044e+03    NaN   
std                 NaN          NaN     NaN        NaN  4.811044e+05    NaN   
min                 NaN          NaN     NaN        NaN  0.000000e+00    NaN   
25%                 NaN          NaN     NaN        NaN  1.100000e+03    NaN   
50%                 NaN          NaN     NaN        NaN  2.950000e+03    NaN   
75%                 NaN          NaN     NaN        NaN  7.200000e+03    NaN   
max                 NaN          NaN     NaN        NaN  1.000000e+08    NaN   

       vehicle_type  registration_year 

0    50000
Name: nr_of_pictures, dtype: int64

We conclude that 'nr_of_pictures' may be dropped since all the values seem to be 0.0 

In [15]:
odo_km = autos.loc[:, 'odometer_km']
odo_km.unique().shape

(13,)

In [16]:
odo_km.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [17]:
odo_km.value_counts().sort_index(ascending = False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [18]:
price_col = autos.loc[:, 'price']

In [19]:
price_col.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
price_col.unique().shape

(2357,)

In [21]:
price_col.value_counts().sort_index(ascending = False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

In [22]:
price_col.value_counts().sort_index(ascending = True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64

In [23]:
price_col.median()

2950.0

In [24]:
price_col.mean()

9840.04376

In [25]:
autos = autos[autos["price"].between(1,351000)]


In [26]:
autos

date_crawled                                               name  \
0      3/26/2016 17:47                   Peugeot_807_160_NAVTECH_ON_BOARD   
1       4/4/2016 13:38         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      3/26/2016 18:57                         Volkswagen_Golf_1.6_United   
3      3/12/2016 16:58  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4       4/1/2016 14:38  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                ...                                                ...   
49995  3/27/2016 14:38   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  3/28/2016 10:50  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997   4/2/2016 14:44                     Fiat_500_C_1.2_Dualogic_Lounge   
49998   3/8/2016 19:25                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999   3/14/2016 0:42                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage      ad_created  nr_of_pictures  \
0         peugeot              nein  3/26/2016 0:00               0   
1             bmw              nein   4/4/2016 0:00               0   
2      volkswagen              nein  3/26/2016 0:00               0   
3           smart              nein  3/12/2016 0:00               0   
4            ford              nein   4/1/2016 0:00               0   
...           ...               ...             ...             ...   
49995        audi              nein  3/27/2016 0:00               0   
49996        opel              nein  3/28/2016 0:00               0   
49997        fiat              nein   4/2/2016 0:00               0   
49998        audi              nein   3/8/2016 0:00               0   
49999        opel              nein  3/13/2016 0:00               0   

       postal_code       last_seen  
0            79588   4/6/2016 6:45  
1            71034  4/6/2016 14:45  
2            35394  4/6/2016 20:15  
3            33729  3/15/2016 3:16  
4            39218  4/1/2016 14:38  
...            ...             ...  
49995        82131  4/1/2016 13:47  
49996        44807  4/2/2016 14:18  
49997        

In [27]:
autos.loc[:, "price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

After we took away some outliers, the median still remains the same as the mean goes down. This may indicate a more normal distribution as there is less skew and the median and mean are closer to one another

In [30]:
autos.loc[:, 'date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

3/10/2016     0.032184
3/11/2016     0.032575
3/12/2016     0.036920
3/13/2016     0.015670
3/14/2016     0.036549
                ...   
4/7/2016 4    0.000021
4/7/2016 6    0.000021
4/7/2016 7    0.000082
4/7/2016 8    0.000082
4/7/2016 9    0.000082
Name: date_crawled, Length: 131, dtype: float64

In [31]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

1/10/2016     0.000041
1/13/2016     0.000021
1/14/2016     0.000021
1/16/2016     0.000021
1/22/2016     0.000021
                ...   
4/6/2016 0    0.003253
4/7/2016 0    0.001256
6/11/2015     0.000021
8/10/2015     0.000021
9/9/2015 0    0.000021
Name: ad_created, Length: 76, dtype: float64

In [32]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

3/10/2016     0.010666
3/11/2016     0.012375
3/12/2016     0.023783
3/13/2016     0.008895
3/14/2016     0.012602
                ...   
4/7/2016 5    0.009575
4/7/2016 6    0.009389
4/7/2016 7    0.009204
4/7/2016 8    0.008875
4/7/2016 9    0.008731
Name: last_seen, Length: 130, dtype: float64

By taking the first 10 values of the string in 'date_crawled', 'ad_created', 'last_seen' we get the date. Based off this, we create a value counts that accounts for null values and results with the values as a percentage. After this, we put it into ascending based off of the dates.  
  
  The date_crawled indicates the time between March and April of 2016 which would allow us to assume that the crawling took time for one single month. The ad_created indicates taht ads were creating mainly between January and April, however there were only a few after the month of April. The last_seen indicates what days the listing was closed or sold and how frequent within those days
   
   

In [33]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration_year indicates that the minimum year was 1000 and max is 9999 which would both be impossible

In [34]:
autos['registration_year'].value_counts().sort_index().tail(30)

2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

In [35]:
autos['registration_year'].value_counts().sort_index().head(30)

1000     1
1001     1
1111     1
1800     2
1910     5
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     4
1957     2
1958     4
1959     6
1960    23
1961     6
1962     4
1963     8
1964    12
Name: registration_year, dtype: int64

We create a bound of 1900 to 2016 as an acceptable range for registration year since cars were invented in 1886, so we can round the start year to 1900. For the end year, the csv/ data was created in 2016 so the end bound must be 2016. 

In [36]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [38]:
autos['registration_year'].value_counts().sort_index().head(20)

1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
1954    2
1955    2
1956    4
1957    2
1958    4
Name: registration_year, dtype: int64

In [39]:
autos['registration_year'].value_counts().sort_index().tail(20)

1997    1951
1998    2363
1999    2897
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
Name: registration_year, dtype: int64

In [41]:
autos['registration_year'].value_counts(normalize=True).sort_index(ascending = False).head(30)

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
2006    0.057197
2005    0.062895
2004    0.057904
2003    0.057818
2002    0.053255
2001    0.056468
2000    0.067608
1999    0.062060
1998    0.050620
1997    0.041794
1996    0.029412
1995    0.026285
1994    0.013474
1993    0.009104
1992    0.007926
1991    0.007262
1990    0.007433
1989    0.003727
1988    0.002892
1987    0.001542
Name: registration_year, dtype: float64

With percentages for the values, most of the cars were registered from 1996 - 2016

In [47]:
autos['brand'].value_counts(normalize = True).head(30)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
Name: brand, dtype: float64

Choose brands that represent more than 2% of the sales, which would mean that Fiat would be last car brand included in our list.

In [60]:
auto_brand_specific = autos['brand'].value_counts(normalize = True).head(9)

In [61]:
auto_brand_specific.index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')

In [62]:
index_brand = auto_brand_specific.index

We are going to create a dictionary with the mean prices of each brand

In [75]:
brand_mean_list = {}

In [76]:
for a in index_brand:
    brand_data = autos.loc[autos['brand'] == a, :]
    mean_of_brand = brand_data['price'].mean()
    brand_mean_list[a] = mean_of_brand

In [80]:
brand_mean_list

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287,
 'renault': 2474.8646069968195,
 'peugeot': 3094.0172290021537,
 'fiat': 2813.748538011696}

For each brand within the Top 9, we have created a dictionary with the mean prices of the cars. We can see that Audi has the most expensive cars and Renault has the least expensive cars in terms of the mean prices. 

In [84]:
new_brand_mean_list = sorted(brand_mean_list.items(), key = lambda kv: kv[1], reverse = True)


In [85]:
new_brand_mean_list

[('audi', 9336.687453600594),
 ('mercedes_benz', 8628.450366422385),
 ('bmw', 8332.820517811953),
 ('volkswagen', 5402.410261610221),
 ('ford', 3749.4695065890287),
 ('peugeot', 3094.0172290021537),
 ('opel', 2975.2419354838707),
 ('fiat', 2813.748538011696),
 ('renault', 2474.8646069968195)]

In [86]:
mean_mileage = {}
for a in index_brand:
    brand_list = autos.loc[autos['brand'] == a, :]
    mileage_mean = brand_list['odometer_km'].mean()
    mean_mileage[a] = mileage_mean

In [92]:
mean_mileage

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558}

In [87]:
mean_price_series = pd.Series(brand_mean_list)

In [89]:
mean_price_dataframe = pd.DataFrame(mean_price_series, columns = ['Mean_Price'])

In [90]:
mean_price_dataframe

Mean_Price
volkswagen     5402.410262
bmw            8332.820518
opel           2975.241935
mercedes_benz  8628.450366
audi           9336.687454
ford           3749.469507
renault        2474.864607
peugeot        3094.017229
fiat           2813.748538

In [91]:
mean_mileage_series = pd.Series(mean_mileage)

In [95]:
mean_price_dataframe['Mean_Mileage'] = mean_mileage_series

In [96]:
mean_price_dataframe

Mean_Price   Mean_Mileage
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872
renault        2474.864607  128071.331213
peugeot        3094.017229  127153.625269
fiat           2813.748538  117121.971596

There does not seem to be a strong correlation between these two in terms of if low mileage would mean a higher price